### PR TITLE
[RuntimeEnv] Reschedule Task when Runtime Env Setup Failed

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

Reschedule task when runtime env setup failed.

When a runtime env is failed to be set up in a node, it will be spill over to another node.

When there is no node left or maximum retry time is reached. This task will be canceled.

## Related issue number

https://github.com/ray-project/ray/issues/16800

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by lixin-wei** [comment]
I'm still working on adding tests and fixing CI. But the code is ready to be reviewed, you can have a look first.

**Comment by edoakes** [comment]
@rkooo567 could you help review please?

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -504,6 +504,9 @@ RAY_CONFIG(bool, scheduler_avoid_gpu_nodes, true)
 /// Whether to skip running local GC in runtime env.
 RAY_CONFIG(bool, runtime_env_skip_local_gc, false)
 
+/// Maximum time to retry task when runtime env setup failed.
+RAY_CONFIG(int32_t, runtime_env_max_retry_time, 2)
```
Comment (edoakes) [inline src/ray/common/ray_config_def.h:508]: specify that this will retry on a different node.

**Review by edoakes — COMMENTED** [review]
@lixin-wei Code looks reasonable to me. Could you also please add some Python integrations for this? You can use `ray.cluster_utils` to mock out a multi-node cluster. I think we have some runtime_env tests that already force failures, would need to modify it to only fail on specific nodes.

**Comment by rkooo567** [comment]
Is this still WIP?

**Comment by rkooo567** [comment]
Btw, when's the case the runtime env setup fails, but we should retry? Can we specify the exact scenario?

**Comment by lixin-wei** [comment]
@rkooo567 All finished, would you please have a look?

**Comment by lixin-wei** [comment]
> Btw, when's the case the runtime env setup fails, but we should retry? Can we specify the exact scenario?

I think there might be nodes with some network issues, and can't download dependencies in time. In this case, we should retry at a different node.

**Comment by lixin-wei** [comment]
@edoakes @rkooo567 All finished, would you please have a look?

**Comment by SongGuyang** [comment]
Please add some description about the reschedule strategy.

File: dashboard/modules/runtime_env/runtime_env_agent.py [hunk 2]
```diff
@@ -84,6 +84,17 @@ def get_or_create_logger(self, job_id: bytes):
...
+        # For testing only
+        node_name = os.environ.get("NODE_NAME")
+        fail_at = os.environ.get("_RAY_RUNTIME_ENV_FAIL_AT_NODE_NAME")
+        logger.debug(f"node_name={node_name}, fail_at={fail_at}")
+        if fail_at == node_name:
+            msg = f"runtime env intentionally failed at node name {node_name}"
+            logger.warning(msg)
+            return runtime_env_agent_pb2.CreateRuntimeEnvReply(
+                status=agent_manager_pb2.AGENT_RPC_STATUS_FAILED,
+                error_message=msg)
```
Comment (edoakes) [inline dashboard/modules/runtime_env/runtime_env_agent.py:96]: can you put this in a utility function please?

**Review by edoakes — APPROVED** [review]
Looks great! Just one small nit :)

**Comment by edoakes** [comment]
Please wait for @rkooo567's review as well

**Comment by edoakes** [comment]
@rkooo567 we want to get this in before branch cut on Monday FYI

**Comment by rkooo567** [comment]
Hmm that’s pretty short timeline for 17 files change PR. Is it absolutely necessary to get this in? I’d like to review carefully and observe the test result if possible

**Comment by rkooo567** [comment]
Just read the description and code, it seems like this is not that urgent PR because

> I think there might be nodes with some network issues, and can't download dependencies in time. In this case, we should retry at a different node.

this scenario is not so common IIUC, and failing tasks/actors with clear error is probably good enough. Wdyt?

**Comment by edoakes** [comment]
Good point @rkooo567, I'm fine w/ pushing it to next release (non-critical for most OSS users)

**Comment by rkooo567** [comment]
So, I have several feedback for the current implementation, but before that why not just naive random retry won't work? Do we need this much of sophisticated mechanism for retry? (blacklisting failed nodes)? 

The failure scenario described here `I think there might be nodes with some network issues, and can't download dependencies in time. In this case, we should retry at a different node.` sounds like most of issues are usually transient?

**Comment by lixin-wei** [comment]
> sounds like most of issues are usually transient?

No in our internal cluster network issue usually takes a long time to recover.

But actually I don't know much about the context of this feature, I was just called to implement this😂

@edoakes @SongGuyang Do you think blacklisting failed node is necessary?

**Comment by SongGuyang** [comment]
For long term, I think we need blacklist. But it is not so easy because we should distinguish local system issues and runtime env resource issues. For local system issues, e.g. network breakdown or local disk exceeded, we should add bad nodes to blacklist and retry to prepare runtime env in another node. But for runtime env resource issues, e.g. the invalid runtime env input or resource not found, we should fail the task immediately and don't reschedule. 

In Ant production, we meet local system issues a lot because we have a large scale shared servers. 

For runtime env, @edoakes Maybe we can subdivide the error code of runtime env install?  For scheduling, in first step, @lixin-wei can we modify the default value of `runtime_env_max_retry_time ` to 1 in current PR? It will keep no-retry by default.  After we finish the error code subdividing, we can integrate the scheduling and add the blacklist strategy.
@rkooo567 What do you think?

**Comment by rkooo567** [comment]
Yeah +1. The current implementation seems to be hard to reason about and cause unexpected retry. I agree the right direction is to implement observability (correct error code) & better error message and implement the blacklist feature afterwards (that’s not tightly coupled to runtime env, but generally working with clean API).

**Comment by rkooo567** [comment]
So the suggestion here is; 

- Implement "better error message" for runtime env first. This will include enum-defined error code for various failures.
- And then implement the scheduling blacklisting 
- And in the last PR, use the blacklist feature based on the error code.

**Comment by SongGuyang** [comment]
@rkooo567 in current PR, @lixin-wei has implemented a task level blacklist. Can we merge current PR first after address all the comments? Or a global blacklist is enough?

**Comment by rkooo567** [comment]
Task level black list == global right? Actor creation == regular task. 

I think we can probably rename some stuff to make sure it is not tightly coupled to runtime env. I will leave some review.

after that, we can set retry == 0, so there’s no behavior change from this pr until we implement further observability

**Comment by lixin-wei** [comment]
> I think we can probably rename some stuff to make sure it is not tightly coupled to runtime env. I will leave some review.

Hi sangbin, what should I rename?

**Comment by rkooo567** [comment]
Sorry for the delay in review. I will review it tomorrow!

**Comment by clarkzinzow** [comment]
@rkooo567 @lixin-wei What's the current status of this? I think we're occasionally hitting transient runtime_env setup failures in our builds, and I assume our users are as well, so this retrying should be a big benefit to both our build system stability and to our users!

**Comment by lixin-wei** [comment]
@clarkzinzow I'm waiting for the review of sangbin, he said some staff should be renamed.

**Comment by rkooo567** [comment]
@clarkzinzow what's the cause of transient failures from the build?

**Comment by clarkzinzow** [comment]
@rkooo567 Unknown, the error message just says that runtime env setup failed.

**Comment by rkooo567** [comment]
Hmm while I will review this pr, in that case we should improve err msg and fix the root cause rather than relying on retry mechanism imo

**Comment by rkooo567** [comment]
Hey @scv119 do you mind reviewing this PR as well? This might introduce a new concept (blacklisting) to the scheduler, so it'd be great to get your feedback

**Comment by clarkzinzow** [comment]
> Hmm while I will review this pr, in that case we should improve err msg and fix the root cause rather than relying on retry mechanism imo

@rkooo567 Yeah I saw a comment from @edoakes on Slack that the error message will be improved I'm the short-term, I was looking to see if that work was ongoing when I found this PR. But it appears to be a transient error that's most likely out of our control (e.g. an external network request failing), so retrying is most likely the correct solution.

> This might introduce a new concept (blacklisting) to the scheduler, so it'd be great to get your feedback

It may be nice to make this a more generic `node_blacklist` concept on the scheduler side, since this should end up being generically useful for future scheduling policies (e.g. anti-affinity/taints + tolerations, as one example).

**Comment by rkooo567** [comment]
Hmm i doubt there are lots of transient network failure in ci. How did you verify that? Also this pr won’t fix the issue if the test is using a single node. My impression is we should do a separate investigation and improve the error message rather than relying on this pr

**Comment by clarkzinzow** [comment]
> Imo i doubt there are lots of transient network failure in ci. How did you verify that?

I wasn't asserting that the runtime env setup failure was due to an external network error, I was just giving that as an example of a transient error. And I'm hypothesizing that the failure is transient since it's not consistently failing in CI (check out `linux://python/ray/tests:test_runtime_env_working_dir_2` on the [flaky test dashboard](https://flakey-tests.ray.io/)).

> Also this pr won’t fix the issue if the test is using a single node.

That's true, I wasn't aware that node blacklisting was part of this PR when I first found it!

> My impression is we should do a separate investigation and improve the error message rather than relying on this pr.

Ah, I think you misunderstood why I pinged on this PR! I just wanted to underline that I think this PR is useful and that we should revive it, since it had been inactive for a few months. I am *not* saying that we should rely on this PR to fix the CI issue. I'm saying that I think this PR will help deal with transient runtime env setup failures, which I _know_ users are encountering since you can find plenty of GitHub issues and Slack posts indicating as much, especially since we'll never be able to eliminate transient failures stemming from external failures like external network requests. I brought up the CI issue since it may be another case of transient runtime env setup failures, but I probably shouldn't have mentioned it since in that case it hasn't been confirmed that we can't fix those transient failures in other ways.

We should obviously fix the error message and drill down on the CI issue, but those efforts are independent of this PR, and I never said that this PR would obviate the need for those efforts.

**Comment by rkooo567** [comment]
ah that makes sense. Thanks for the clarification! (sorry for misunderstanding the comment) Yeah I think this can help resolving some transient failures.

**Comment by jovany-wang** [comment]
Hi @rkooo567 do you have more comments? We're  going to merge this.

**Comment by SongGuyang** [comment]
@lixin-wei @rkooo567 What the status of current PR? Can we merge master and do the final review?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Please do not merge this. I promise to review this PR today.

File: src/ray/common/task/task.cc [hunk 3]
```diff
@@ -37,6 +37,8 @@ const TaskExecutionSpecification &RayTask::GetTaskExecutionSpec() const {
 
 const TaskSpecification &RayTask::GetTaskSpecification() const { return task_spec_; }
 
+TaskSpecification &RayTask::GetMutableTaskSpecification() { return task_spec_; }
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/common/task/task.cc:40]: Also accessing mutable fields should hold a lock

[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/common/task/task.cc:40]: Maybe consider 

```
const void AddBlackListedNode

const void RemoveBlackListedNode

const vector<NodeID> GetBlacklistedNodes const
```

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 4]
```diff
@@ -518,7 +523,15 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
   // same TaskID to request a worker
   auto resource_spec_msg = scheduling_key_entry.resource_spec.GetMutableMessage();
   resource_spec_msg.set_task_id(TaskID::FromRandom(job_id_).Binary());
-  const TaskSpecification resource_spec = TaskSpecification(resource_spec_msg);
+  TaskSpecification resource_spec = TaskSpecification(resource_spec_msg);
+
+  // Copy runtime env rescheduling context.
+  if (reply_last_time != nullptr) {
+    resource_spec.GetMutableMessage()
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_task_transport.cc:530]: This is awkward to be updated in this function. Can we update it before calling this method? This must be updated within the callback of `lease_client->RequestWorkerLease`

File: src/ray/protobuf/common.proto [hunk 5]
```diff
@@ -232,6 +232,9 @@ message TaskSpec {
   // The depth of the task. The driver has depth 0, anything it calls has depth
   // 1, etc.
   int64 depth = 26;
+  // Node IDs that runtime env setup has ever failed at.
+  // These nodes will be skipped when rescheduling.
+  repeated bytes runtime_env_setup_failed_node_ids = 27;
```
Comment (rkooo567) [inline src/ray/protobuf/common.proto:237]: Consider adding a mutable context and add a blacklisted_nodes field

File: src/ray/protobuf/common.proto [hunk 6]
```diff
@@ -232,6 +232,9 @@ message TaskSpec {
   // The depth of the task. The driver has depth 0, anything it calls has depth
   // 1, etc.
   int64 depth = 26;
+  // Node IDs that runtime env setup has ever failed at.
```
Comment (rkooo567) [inline src/ray/protobuf/common.proto:235]:
```
Blacklisted_node_ids. Blacklisted node ids will be excluded from scheduling decision.
```

File: src/ray/protobuf/node_manager.proto [hunk 7]
```diff
@@ -70,6 +70,9 @@ message RequestWorkerLeaseReply {
   bool rejected = 7;
   // The (normal task) resources data to be carried by the Reply.
   ResourcesData resources_data = 8;
+  // Node IDs that runtime env setup has ever failed at.
+  // These nodes will be skipped when rescheduling.
+  repeated bytes runtime_env_setup_failed_node_ids = 9;
```
Comment (rkooo567) [inline src/ray/protobuf/node_manager.proto:75]: same comment as other parts

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 8]
```diff
@@ -212,7 +218,7 @@ bool ClusterResourceScheduler::IsSchedulable(const ResourceRequest &resource_req
 
 int64_t ClusterResourceScheduler::GetBestSchedulableNode(
     const ResourceRequest &resource_request, bool actor_creation, bool force_spillback,
-    int64_t *total_violations, bool *is_infeasible) {
+    int64_t *total_violations, bool *is_infeasible, const TaskSpecification &task_spec) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:221]: Task spec shouldn't be exposed to cluster_resource_scheduler. Can you just pass blacklisted nodes directly here?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 9]
```diff
@@ -264,11 +272,12 @@ int64_t ClusterResourceScheduler::GetBestSchedulableNode(
 std::string ClusterResourceScheduler::GetBestSchedulableNode(
     const absl::flat_hash_map<std::string, double> &task_resources,
     bool requires_object_store_memory, bool actor_creation, bool force_spillback,
-    int64_t *total_violations, bool *is_infeasible) {
+    int64_t *total_violations, bool *is_infeasible, const TaskSpecification &task_spec) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:275]: same comment as above

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 10]
```diff
@@ -37,6 +37,12 @@ namespace ray {
 
 using rpc::HeartbeatTableData;
 
+/// Detailed context when scheduling a task, could contains info about
+/// task/nodes/resources, used for additional scheduling policy.
+struct SchedulingContext {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:42]: This seems to be unnecessary wrapper

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 11]
```diff
@@ -394,6 +401,7 @@ class ClusterResourceScheduler : public ClusterResourceSchedulerInterface {
 
  private:
   bool NodeAlive(int64_t node_id) const;
+  bool IsNodeAvailable(int64_t node_id, const SchedulingContext &context) const;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:404]: Consider removing NodeAlive and replace to this method?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 12]
```diff
@@ -56,6 +56,18 @@ ClusterTaskManager::ClusterTaskManager(
       metric_tasks_dispatched_(0),
       metric_tasks_spilled_(0) {}
 
+void ReplyCancelled(const std::shared_ptr<internal::Work> &work,
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:59]: Is there any other method that we can reuse? I believe we have other logic that is about cancelling lease request

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 13]
```diff
@@ -216,31 +242,54 @@ bool ClusterTaskManager::PoppedWorkerHandler(
...
+            task_dependency_manager_.RemoveTaskDependencies(
+                task.GetTaskSpecification().TaskId());
+          }
         } else {
-          RAY_LOG(FATAL) << "Unexpected state received for the empty pop worker. Status: "
-                         << status;
+          RAY_LOG(INFO) << "Retry runtime env setup, task_id="
+                        << work->task.GetTaskSpecification().TaskId();
+          /// Retry scheduling.
+          work->SetStateWaiting(cause);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:270]:
```
          work->SetStateWaiting(cause);
          erase_from_dispatch_queue_fn(work, scheduling_class);
          tasks_to_schedule_[scheduling_class].push_back(work);
          ScheduleAndDispatchTasks();
```
This seems to be unrelated to this PR? is it a separate bug?

File: src/ray/common/ray_config_def.h [hunk 14]
```diff
@@ -504,6 +504,11 @@ RAY_CONFIG(bool, scheduler_avoid_gpu_nodes, true)
 /// Whether to skip running local GC in runtime env.
 RAY_CONFIG(bool, runtime_env_skip_local_gc, false)
 
+/// Maximum time to retry task when runtime env setup failed.
+/// Task will be retry at a different node, if no feasible node can be found, task will be
+/// canceled immediately regardless of the remaining retry time.
+RAY_CONFIG(int32_t, runtime_env_max_retry_time, 2)
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:510]: Can you set this 0 by default?

File: python/ray/tests/test_runtime_env.py [hunk 15]
```diff
@@ -286,6 +287,76 @@ def f():
         ray.get(a.ready.remote())
 
 
+def test_runtime_env_rescheduling(shutdown_only):
```
Comment (rkooo567) [inline python/ray/tests/test_runtime_env.py:290]: Can you make sure all these scenarios are tested? (please verify you did it!)

- Regular rescheduling (failed once and make sure the task is scheduled on other node).
- Rescheduling failed because every node is blacklisted -> should raise exception
- Rescheduling failed because the only node that has available resources is blacklisted -> should raise exception

File: python/ray/tests/test_runtime_env.py [hunk 16]
```diff
@@ -286,6 +287,76 @@ def f():
...
+    cluster.add_node(env_vars={"NODE_NAME": "1"})
+    cluster.add_node(env_vars={"NODE_NAME": "2"})
+    cluster.connect()
+
+    # test task
+    for i in range(10):
+        node_name = ray.get(
+            get_node_name.options(runtime_env=runtime_env).remote())
+        assert node_name == "2"
+    # test actor
```
Comment (rkooo567) [inline python/ray/tests/test_runtime_env.py:317]: Btw, how does it work without modifying actor manager code? I thought we should modify task spec for actor creation task within gcs?

**Review by rkooo567 — COMMENTED** [review]
Btw, we should empty the blacklist when task is reconstructed. The current implementation assumes blacklisted nodes are permanently unavailable, but this can be different if they are reconstructed.

**Comment by rkooo567** [comment]
@stephanie-wang can you take a look when you have the time?

File: src/ray/protobuf/common.proto (hunk 6 above)
Comment (jovany-wang) [reply src/ray/protobuf/common.proto:235]: I'm not sure if this impacts the perf?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 17]
```diff
@@ -107,10 +119,24 @@ bool ClusterTaskManager::SchedulePendingTasks() {
...
+      // If this task's runtime env has failed to setup in this node, and can't find
+      // another node to retry, cancel this task immediately.
+      if (task_spec.HasRuntimeEnvFailureAtNode(self_node_id_.Binary())) {
+        // Cancel this task directly and raise a `RuntimeEnvSetupError` exception to
+        // user eventually. The task will be removed from dispatch queue in
+        // `CancelTask`.
+        RAY_LOG(WARNING)
+            << "Task " << task_spec.TaskId()
+            << " has been canceled because runtime env setup failed in this node, and "
+               "there isn't any other node feasible.";
```
Comment (jovany-wang) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:133]: Why there is no more node feasible?

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 12 above)
Comment (jovany-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:59]: I believe this isn't a new method.

**Review by jovany-wang — COMMENTED** [review]
At high level, I'm not sure if we should turn this on by default. If we do so, it may slow down the task scheduling when this failure happens.
But this is able to improve the stability indeed when env issues happen.
@ericl Thoughts?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by stephanie-wang** [comment]
What's the status of this PR?

Also, I'd like to understand the motivation for this feature more. What are the cases where the runtime env would fail to be set up on one node but succeeds on another? It seems like a pretty limited scenario to me, and the code to do this is pretty complex (involves adding state to the TaskSpec and changing the scheduling logic). I'm against merging this unless we can: (1) justify that this case happens often enough, and (2) reduce complexity of the design, e.g., keep the tasks stateless.

What's wrong with retrying on the same node and allowing the current scheduling policy to work as usual?

**Comment by rkooo567** [comment]
@stephanie-wang actually I asked the same question before. Doesn’t mean that it should convince you, but for the sake of communication, I’d like to give you the link to the conversation; https://github.com/ray-project/ray/pull/20188#issuecomment-968241998

**Comment by stephanie-wang** [comment]
> @stephanie-wang actually I asked the same question before. Doesn’t mean that it should convince you, but for the sake of communication, I’d like to give you the link to the conversation; [#20188 (comment)](https://github.com/ray-project/ray/pull/20188#issuecomment-968241998)

Ah thanks, this is helpful.

I am not convinced by this context though. You could make the same arguments about tasks that fail to complete because they require the network or local disk. Shall we close this PR until we can come up with a better solution?

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```
